### PR TITLE
Index search

### LINKED ISSUE #412: Search & Import Contact from Index Server
```GHERKIN
As a Qabel Desktop User
In order to write messages to people whose email-address or phonenumber I know
I need to be able to search and import users from the qabel index server
```

In the mocks, we currently only have the menu entry:
image

//TBD with @thechauffeur for more details / scope

### PR DESCRIPTION
**Pull request by julianseeger**
resolves #412 

If the Icons class get's approved, I'm going to convert all existing Image- and ImageView-Usages to that class in a separate PR.

### CODE REVIEW

File: launch.properties [hunk 1]
```diff
@@ -0,0 +1,5 @@
+drop.url=https://drop.qabel.org
+accounting.url=https://accounting.qabel.org
+block.url=https://block.qabel.org
+index.url=https://index.qabel.org
+crashReport.url=https://rink.hockeyapp.net/api/2/apps/
```
Comment (audax) [inline launch.properties]: Did you mean to commit this?

File: src/main/java/de/qabel/desktop/ui/contact/AssignContactPopover.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+public class AssignContactPopover extends PopOver {
+    public AssignContactPopover(Contact contact) {
+        super(new AssignContactView(contact).getView());
+
+        setTitle(contact.getAlias());
+        setHeaderAlwaysVisible(true);
+        getStyleClass().add("assignContactPopover");
+        getRoot().getStyleClass().add("assignContactPopover");
+        getRoot().getStylesheets().add(QabelFXMLView.getGlobalStyleCheat());
```
Comment (audax) [inline src/main/java/de/qabel/desktop/ui/contact/AssignContactPopover.java]: StyleCheat? Not StyleSheet? Oo

File: src/main/java/de/qabel/desktop/ui/contact/context/AssignContactController.java [hunk 3]
```diff
@@ -63,6 +73,18 @@ public void initialize(URL location, ResourceBundle resources) {
...
+        char[] chars = keyIdentifier.toCharArray();
+        StringBuilder formatted = new StringBuilder();
+        for (int i = 0; i < chars.length; i++) {
+            if (i > 0 && i%4 == 0) {
+                formatted.append(" ");
+            }
+            formatted.append(chars[i]);
+        }
+        return formatted.toString();
+    }
```
Comment (audax) [inline src/main/java/de/qabel/desktop/ui/contact/context/AssignContactController.java]: I think we have a method in the core for that. Maybe in the form of an extension method, so you maybe have to wrap it.

File: src/main/java/de/qabel/desktop/ui/contact/index/IndexSearchController.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    private void decrementPendingRequests() {
+        synchronized (pendingRequests) {
+            pendingRequests.set(pendingRequests.get()-1);
+        }
+    }
+
+    private List<Contact> search(String searchText) {
+        incrementPendingRequests();
```
Comment (audax) [inline src/main/java/de/qabel/desktop/ui/contact/index/IndexSearchController.java]: Couldn't you just cancel other pending requests? Is there a need to keep "old" requests around?

File: src/main/java/de/qabel/desktop/ui/util/Icons.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
+package de.qabel.desktop.ui.util;
+
+import javafx.scene.image.Image;
+import javafx.scene.image.ImageView;
+
+public class Icons {
```
Comment (audax) [inline src/main/java/de/qabel/desktop/ui/util/Icons.java]: Looks fine to me, I don't care much if the icon paths live in some xml/properties-file or directly as statics in the class.

File: src/main/java/de/qabel/desktop/ui/contact/index/IndexSearchController.java (hunk 4 above)
Comment (julianseeger) [inline src/main/java/de/qabel/desktop/ui/contact/index/IndexSearchController.java]: It's just for the spinner o.O

File: launch.properties (hunk 1 above)
Comment (julianseeger) [inline launch.properties]: :-X

File: src/main/resources/ui.properties [hunk 6]
```diff
@@ -216,3 +216,5 @@ saveAndUploadIdentity=Save & Publish
 emailLabel=Email
 phoneLabel=Phone
 aliasLabel=Alias
+indexSearchPrompt=Enter full E-Mail-Address or Phonenumber
+assignContactTitle=assign to identities
```
Comment (audax) [inline src/main/resources/ui.properties]: Messages from android:
https://github.com/Qabel/qabel-android/pull/662/files#diff-a10862e056a49e534adc98c5b2cf6676R360

File: src/main/java/de/qabel/desktop/ui/contact/context/AssignContactController.java (hunk 3 above)
Comment (julianseeger) [inline src/main/java/de/qabel/desktop/ui/contact/context/AssignContactController.java]: Already searched for that but couldn't find any.... hints appreciated ;)

File: src/main/java/de/qabel/desktop/ui/contact/index/IndexSearchController.java (hunk 4 above)
Comment (julianseeger) [inline src/main/java/de/qabel/desktop/ui/contact/index/IndexSearchController.java]: https://github.com/Qabel/qabel-desktop/pull/461/files#diff-b849be8f2cb2ad0a1b83d5b2cc55bf19R54

Comment (audax) [inline src/main/java/de/qabel/desktop/ui/contact/index/IndexSearchController.java]: Can "pendingRequests" be greater than 1?

File: launch.properties (hunk 1 above)
Comment (julianseeger) [inline launch.properties]: ✓

File: src/main/java/de/qabel/desktop/ui/contact/AssignContactPopover.java (hunk 2 above)
Comment (julianseeger) [inline src/main/java/de/qabel/desktop/ui/contact/AssignContactPopover.java]: ✓

File: src/main/resources/ui.properties (hunk 6 above)
Comment (julianseeger) [inline src/main/resources/ui.properties]: ✓

File: src/main/java/de/qabel/desktop/ui/contact/index/IndexSearchController.java (hunk 4 above)
Comment (julianseeger) [inline src/main/java/de/qabel/desktop/ui/contact/index/IndexSearchController.java]: ftr.: yes, if the first search takes more than the debounceTimout=500ms, then a second searchText could arrive while the first is still running, increasing the pendingRequests to 2

**Pull request merged**